ykrt/src/compile/jitc_yk/arbbitint.rs

```diff
@@ -0,0 +1,840 @@
+        Self {
+            bitw: self.bitw,
+            val: self.val.wrapping_mul(other.val),
+        }
+    }
+
+    /// Return a new [ArbBitInt] that performs two's complement wrapping subtraction on `self` and
+    /// `other`.
+    ///
+    /// # Panics
+    ///
+    /// If `self` and `other` are not the same bit width.
+    pub(crate) fn wrapping_sub(&self, other: &Self) -> Self {
+        debug_assert_eq!(self.bitw, other.bitw);
+        Self {
+            bitw: self.bitw,
+            val: self.val.wrapping_sub(other.val),
+        }
+    }
+
+    /// Return a new [ArbBitInt] that left shifts `self` by `bits`s or `None` if `bits >=
+    /// self.bitw()`.
+    pub(crate) fn checked_shl(&self, bits: u32) -> Option<Self> {
+        if bits < self.bitw {
+            Some(Self {
+                bitw: self.bitw,
+                val: self.val.checked_shl(bits).unwrap(), // unwrap cannot fail
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Return a new [ArbBitInt] that right shifts `self` by `bits` or `None` if `bits >=
+    /// self.bitw()`.
+    pub(crate) fn checked_shr(&self, bits: u32) -> Option<Self> {
+        if bits < self.bitw {
+            Some(Self {
+                bitw: self.bitw,
+                val: self.val.checked_shr(bits).unwrap(), // unwrap cannot fail
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Return a new [ArbBitInt] that performs bitwise `AND` on `self` and `other`.
+    ///
+    /// # Panics
+    ///
+    /// If `self` and `other` are not the same bit width.
+    pub(crate) fn bitand(&self, other: &Self) -> Self {
+        debug_assert_eq!(self.bitw, other.bitw);
+        Self {
+            bitw: self.bitw,
+            val: self.val.bitand(other.val),
+        }
+    }
+
+    /// Return a new [ArbBitInt] that performs bitwise `OR` on `self` and `other`.
+    ///
+    /// # Panics
+    ///
+    /// If `self` and `other` are not the same bit width.
+    pub(crate) fn bitor(&self, other: &Self) -> Self {
+        debug_assert_eq!(self.bitw, other.bitw);
+        Self {
+            bitw: self.bitw,
+            val: self.val.bitor(other.val),
+        }
+    }
+
+    /// Return a new [ArbBitInt] that performs bitwise `XOR` on `self` and `other`.
+    ///
+    /// # Panics
+    ///
+    /// If `self` and `other` are not the same bit width.
+    pub(crate) fn bitxor(&self, other: &Self) -> Self {
+        debug_assert_eq!(self.bitw, other.bitw);
+        Self {
+            bitw: self.bitw,
+            val: self.val.bitxor(other.val),
+        }
+    }
+}
+
+impl Hash for ArbBitInt {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.bitw.hash(state);
+        self.val.truncate(self.bitw).hash(state);
+    }
+}
+
+impl PartialEq for ArbBitInt {
+    fn eq(&self, other: &Self) -> bool {
+        self.bitw == other.bitw && self.val.truncate(self.bitw) == other.val.truncate(self.bitw)
```

Comment:
I wonder if comparing different bit sizes should panic here?

Reply:
I think because we can hash `Const`s that wouldn't be the right thing to do here: we will end up comparing different bit sized things.

Reply:
Ah, right.